tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/ClientCredentialsTests.cs

```diff
@@ -246,30 +248,31 @@ private static IConfidentialClientApplication CreateApp(CredentialType credentia
         private string GetExpectedCacheKey(string clientId, string tenantId)
         {
             return $"{clientId}_{tenantId ?? ""}_AppTokenCache";
-        }       
+        }
 
         internal static long ConvertToTimeT(DateTime time)
         {
             var startTime = new DateTime(1970, 1, 1, 0, 0, 0, 0);
             TimeSpan diff = time - startTime;
             return (long)diff.TotalSeconds;
         }
-
         private static IDictionary<string, string> GetClaims(bool useDefaultClaims = true)
         {
+            const uint JwtToAadLifetimeInSeconds = 60 * 10; // Ten minutes
+
             if (useDefaultClaims)
             {
-                DateTime validFrom = DateTime.UtcNow;
-                var nbf = ConvertToTimeT(validFrom);
-                var exp = ConvertToTimeT(validFrom + TimeSpan.FromSeconds(TestConstants.JwtToAadLifetimeInSeconds));
+                DateTimeOffset now = DateTimeOffset.UtcNow;
+                DateTimeOffset validFrom = now; // AAD will take clock skew into consideration
+                DateTimeOffset validUntil = now.AddSeconds(JwtToAadLifetimeInSeconds);
 
                 return new Dictionary<string, string>()
                 {
                 { "aud", TestConstants.ClientCredentialAudience },
-                { "exp", exp.ToString(CultureInfo.InvariantCulture) },
+                { "exp", validUntil.ToUnixTimeSeconds().ToString(CultureInfo.InvariantCulture) },
```

Comment:
Setting the `exp` or `nbf` to a string will fail to generate a token on the actual login server.

Reply:
This is not my experience, probably AAD is resilient to this, but I agree it's better to respect the types and not force strings.